odxtools/diaglayerraw.py

```diff
@@ -0,0 +1,283 @@
+
+        parent_refs = [
+            ParentRef.from_et(pr_el, doc_frags)
+            for pr_el in et_element.iterfind("PARENT-REFS/PARENT-REF")
+        ]
+
+        communication_parameters = [
+            CommunicationParameterRef.from_et(el, doc_frags, variant_type)
+            for el in et_element.iterfind("COMPARAM-REFS/COMPARAM-REF")
+        ]
+
+        ecu_variant_patterns = [
+            EcuVariantPattern.from_et(el, doc_frags)
+            for el in et_element.iterfind("ECU-VARIANT-PATTERN/ECU-VARIANT-PATTERN")
+        ]
+        if variant_type is not DiagLayerType.ECU_VARIANT:
+            assert (
+                len(ecu_variant_patterns) == 0
+            ), "DiagLayer of type other than 'ECU-VARIANT' must not define a ECU-VARIANT-PATTERN"
+
+        # Create DiagLayer
+        return DiagLayerRaw(
+            variant_type=variant_type,
+            odx_id=odx_id,
+            short_name=short_name,
+            long_name=long_name,
+            description=description,
+            admin_data=admin_data,
+            company_datas=NamedItemList(short_name_as_id, company_datas),
+            functional_classes=NamedItemList(short_name_as_id, functional_classes),
+            diag_data_dictionary_spec=diag_data_dictionary_spec,
+            diag_comms=diag_comms,
+            requests=NamedItemList(short_name_as_id, requests),
+            positive_responses=NamedItemList(short_name_as_id, positive_responses),
+            negative_responses=NamedItemList(short_name_as_id, negative_responses),
+            global_negative_responses=NamedItemList(short_name_as_id, global_negative_responses),
+            import_refs=import_refs,
+            state_charts=NamedItemList(short_name_as_id, state_charts),
+            additional_audiences=NamedItemList(short_name_as_id, additional_audiences),
+            sdgs=sdgs,
+            parent_refs=parent_refs,
+            communication_parameters=communication_parameters,
+            ecu_variant_patterns=ecu_variant_patterns,
+        )
+
+    def _build_odxlinks(self) -> Dict[OdxLinkId, Any]:
+        """Construct a mapping from IDs to all objects that are contained in this diagnostic layer."""
+        odxlinks = {self.odx_id: self}
+
+        for obj in chain(
+            [self.admin_data],
+                self.company_datas,
+                self.functional_classes,
+            [self.diag_data_dictionary_spec],
+                self.diag_comms,
+                self.requests,
+                self.positive_responses,
+                self.negative_responses,
+                self.global_negative_responses,
+                self.state_charts,
+                self.additional_audiences,
+                self.sdgs,
+                self.parent_refs,
+                self.communication_parameters,
+        ):
+            if obj is None or isinstance(obj, OdxLinkRef):
+                continue
+
+            if hasattr(obj, "odx_id"):
+                odxlinks[obj.odx_id] = obj
```

Comment:
I don't get this part, what are you doing here?

Reply:
some objects here (erhm, `diag_data_dictionary_spec`) do not have an `odx_id`,  but contain objects which have it, so their `_build_odxlinks()` method needs to be called. This issue can be avoided if it is guaranteed that all objects add themselfs in `_build_odxlinks()` if they exhibit `.odx_id`. I think almost all do, but I did not have the nerves to go hunting for the ones which don't. I'll have a look...

Reply:
so this is a hack that needs to be fixed? maybe add TODO?

Reply:
> I'll have a look...

Okay, done. the result is pretty large and I don't think this should be part of the present PR. I'll submit that cleanup once this PR is in...